File: frontend/src/types/reports.ts
// Report Module Types

export type ReportCategory = 'operational' | 'financial' | 'guest' | 'performance' | 'inventory' | 'compliance' | 'custom';
export type ReportType = 'standard' | 'custom' | 'dashboard' | 'scheduled';
export type ReportFrequency = 'daily' | 'weekly' | 'monthly' | 'quarterly' | 'yearly' | 'custom';
export type ReportFormat = 'pdf' | 'excel' | 'csv' | 'json';
export type AlertSeverity = 'info' | 'warning' | 'critical';

// Report Definition
export interface ReportDefinition {
  id: string;
  report_code: string;
  report_name: string;
  report_category: ReportCategory;
  report_type: ReportType;
  description?: string;
  parameters?: ReportParameters;
  display_config?: DisplayConfig;
  required_permission?: string;
  visibility: 'public' | 'private' | 'department' | 'role';
  is_active: boolean;
  is_system: boolean;
  created_at: string;
  updated_at: string;
}

export interface ReportParameters {
  [key: string]: {
    type: 'date' | 'daterange' | 'select' | 'multiselect' | 'text' | 'number' | 'month';
    required: boolean;
    default?: any;
    options?: string[];
  };
}

export interface DisplayConfig {
  charts?: string[];
  tables?: boolean;
  summary_cards?: boolean;
  export_formats?: ReportFormat[];
  real_time?: boolean;
  refresh_interval?: number;
}

// Report Schedule
export interface ReportSchedule {
  id: string;
  report_id: string;
  schedule_name: string;
  frequency: ReportFrequency;
  schedule_config: ScheduleConfig;
  default_parameters?: Record<string, any>;
  recipients: Recipients;
  delivery_format: ReportFormat;
  last_run_at?: string;
  next_run_at?: string;
  last_status?: 'pending' | 'running' | 'completed' | 'failed';
  error_message?: string;
  is_active: boolean;
  created_at: string;
}

export interface ScheduleConfig {
  time?: string;
  day_of_week?: number;
  day_of_month?: number;
  cron_expression?: string;
}

export interface Recipients {
  emails?: string[];
  roles?: string[];
  slack_channels?: string[];
}

// Report Execution
export interface ReportExecution {
  id: string;
  report_id: string;
  schedule_id?: string;
  executed_by: string;
  execution_type: 'manual' | 'scheduled' | 'api';
  parameters: Record<string, any>;
  row_count?: number;
  execution_time_ms?: number;
  file_url?: string;
  file_size?: number;
  status: 'pending' | 'running' | 'completed' | 'failed';
  error_message?: string;
  executed_at: string;
  completed_at?: string;
}

// KPI Definition
export interface KPIDefinition {
  id: string;
  kpi_code: string;
  kpi_name: string;
  category?: string;
  calculation_sql?: string;
  unit: 'percentage' | 'currency' | 'number' | 'ratio';
  target_value?: number;
  warning_threshold?: number;
  critical_threshold?: number;
  display_format?: string;
  trend_direction?: 'higher_better' | 'lower_better' | 'neutral';
  is_active: boolean;
  created_at: string;
}

export interface KPIValue {
  id: string;
  kpi_id: string;
  value: number;
  period_date: string;
  period_type: 'daily' | 'weekly' | 'monthly';
  previous_value?: number;
  change_percentage?: number;
  status?: 'good' | 'warning' | 'critical';
  calculated_at: string;
}

// Report Data Structures

// Daily Operations Report
export interface DailyOperationsReport {
  report_date: string;
  summary: {
    occupancy: OccupancySummary;
    revenue: RevenueSummary;
    operations: OperationsSummary;
  };
  detailed_sections: {
    arrivals: ArrivalData[];
    departures: DepartureData[];
    housekeeping_status: HousekeepingStatus;
    alerts: OperationalAlert[];
  };
}

export interface OccupancySummary {
  total_rooms: number;
  occupied_rooms: number;
  occupancy_rate: number;
  arrivals_today: number;
  departures_today: number;
  in_house_guests?: number;
}

export interface RevenueSummary {
  room_revenue: number;
  pos_revenue: number;
  other_revenue?: number;
  total_revenue: number;
  adr: number;
  rev_par: number;
}

export interface OperationsSummary {
  check_ins_completed: number;
  check_outs_completed: number;
  pending_arrivals: number;
  pending_departures: number;
  rooms_to_clean: number;
  maintenance_issues: number;
}

export interface ArrivalData {
  booking_code: string;
  guest_name: string;
  room: string;
  check_in_time: string;
  nights: number;
  status: string;
  special_requests?: string;
}

export interface DepartureData {
  booking_code: string;
  guest_name: string;
  room: string;
  check_out_time: string;
  balance_due?: number;
}

export interface HousekeepingStatus {
  clean: number;
  dirty: number;
  inspected: number;
  out_of_order: number;
}

export interface OperationalAlert {
  type: string;
  severity: AlertSeverity;
  message: string;
}

// Revenue Report
export interface RevenueReport {
  summary: {
    total_revenue: number;
    room_revenue: number;
    fnb_revenue: number;
    other_revenue: number;
    growth_rate: number;
    vs_last_period: string;
  };
  revenue_breakdown: {
    by_source: Record<string, number>;
    by_room_type: Record<string, number>;
    by_payment_method: Record<string, number>;
  };
  daily_trend?: TrendData[];
  top_performers?: {
    best_room_type: string;
    best_day: string;
    highest_adr_day: string;
  };
  year_over_year?: {
    current_year: number;
    last_year: number;
    growth: number;
    growth_percentage: number;
  };
}

export interface TrendData {
  date: string;
  value: number;
  label?: string;
}

// Profit & Loss Report
export interface ProfitLossReport {
  period: string;
  revenue: {
    room_revenue: number;
    food_beverage: number;
    other_operating: number;
    total_revenue: number;
  };
  expenses: {
    operating_expenses: Record<string, number>;
    non_operating_expenses: Record<string, number>;
    total_expenses: number;
  };
  profitability: {
    gross_profit: number;
    operating_profit: number;
    net_profit: number;
    profit_margin: number;
    ebitda: number;
  };
  ratios: {
    expense_ratio: number;
    labor_cost_ratio: number;
    gop_margin: number;
  };
  comparison?: {
    vs_last_month: string;
    vs_last_year: string;
    vs_budget: string;
  };
}

// Guest Demographics Report
export interface GuestDemographicsReport {
  period: string;
  total_guests: number;
  demographics: {
    by_nationality: Record<string, number>;
    by_age_group: Record<string, number>;
    by_gender: Record<string, number>;
    by_purpose: Record<string, number>;
  };
  behavior_patterns: {
    average_length_of_stay: number;
    booking_lead_time: number;
    repeat_guest_rate: number;
    direct_booking_rate: number;
  };
  preferences: {
    room_type: Record<string, number>;
    payment_method: Record<string, number>;
  };
}

// KPI Dashboard
export interface KPIDashboard {
  timestamp: string;
  kpis: Record<string, KPIData>;
  alerts: KPIAlert[];
  period_comparison: {
    vs_yesterday?: string;
    vs_last_week?: string;
    vs_last_month?: string;
    vs_last_year?: string;
  };
}

export interface KPIData {
  current: number;
  target: number;
  trend: 'up' | 'down' | 'stable';
  status: 'good' | 'warning' | 'critical' | 'excellent';
}

export interface KPIAlert {
  kpi: string;
  message: string;
  severity: AlertSeverity;
}

// Report Filters
export interface SavedFilter {
  id: string;
  report_id: string;
  user_id: string;
  filter_name: string;
  filter_config: Record<string, any>;
  is_default: boolean;
  created_at: string;
}

// Report Subscription
export interface ReportSubscription {
  id: string;
  user_id: string;
  report_id: string;
  frequency?: ReportFrequency;
  delivery_method: 'email' | 'in_app' | 'slack';
  format: ReportFormat;
  custom_filters?: Record<string, any>;
  is_active: boolean;
  subscribed_at: string;
}

// Chart Configuration
export interface ChartConfig {
  type: 'line' | 'bar' | 'pie' | 'donut' | 'area' | 'scatter' | 'heatmap' | 'gauge' | 'radar';
  title?: string;
  data: any[];
  xAxis?: string;
  yAxis?: string;
  series?: ChartSeries[];
  colors?: string[];
  legend?: boolean;
  tooltip?: boolean;
  responsive?: boolean;
}

export interface ChartSeries {
  name: string;
  data: number[];
  type?: string;
  color?: string;
}

// Export Options
export interface ExportOptions {
  format: ReportFormat;
  include_charts?: boolean;
  include_summary?: boolean;
  page_orientation?: 'portrait' | 'landscape';
  paper_size?: 'A4' | 'Letter' | 'Legal';
}

// Report Request/Response
export interface ExecuteReportRequest {
  report_id: string;
  parameters: Record<string, any>;
  format?: ReportFormat;
  include_charts?: boolean;
}

export interface ExecuteReportResponse {
  success: boolean;
  data?: any;
  file_url?: string;
  error?: string;
  execution_id?: string;
}

export interface CreateScheduleRequest {
  report_id: string;
  schedule_name: string;
  frequency: ReportFrequency;
  schedule_config: ScheduleConfig;
  default_parameters?: Record<string, any>;
  recipients: Recipients;
  delivery_format: ReportFormat;
}

File: backend/app/schemas/reports.py
from datetime import date, datetime
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, Field
from decimal import Decimal
from enum import Enum
from uuid import UUID


class ReportPeriod(str, Enum):
    DAILY = "daily"
    WEEKLY = "weekly"
    MONTHLY = "monthly"
    QUARTERLY = "quarterly"
    YEARLY = "yearly"
    CUSTOM = "custom"


class ReportCategory(str, Enum):
    OPERATIONAL = "operational"
    FINANCIAL = "financial"
    GUEST = "guest"
    PERFORMANCE = "performance"
    INVENTORY = "inventory"
    COMPLIANCE = "compliance"
    CUSTOM = "custom"


class ReportType(str, Enum):
    STANDARD = "standard"
    CUSTOM = "custom"
    DASHBOARD = "dashboard"
    SCHEDULED = "scheduled"


class ReportFormat(str, Enum):
    PDF = "pdf"
    EXCEL = "excel"
    CSV = "csv"
    JSON = "json"


class AlertSeverity(str, Enum):
    INFO = "info"
    WARNING = "warning"
    CRITICAL = "critical"


class ReportDefinition(BaseModel):
    id: UUID
    report_code: str
    report_name: str
    report_category: ReportCategory
    report_type: ReportType
    description: Optional[str] = None
    parameters: Optional[Dict[str, Any]] = None
    display_config: Optional[Dict[str, Any]] = None
    required_permission: Optional[str] = None
    visibility: str = "public"
    is_active: bool = True
    is_system: bool = False
    created_at: datetime
    updated_at: datetime


class ReportSchedule(BaseModel):
    id: UUID
    report_id: UUID
    schedule_name: str
    frequency: ReportPeriod
    schedule_config: Dict[str, Any]
    default_parameters: Optional[Dict[str, Any]] = None
    recipients: Dict[str, Any]
    delivery_format: ReportFormat
    last_run_at: Optional[datetime] = None
    next_run_at: Optional[datetime] = None
    last_status: Optional[str] = None
    error_message: Optional[str] = None
    is_active: bool = True
    created_at: datetime


class ReportExecution(BaseModel):
    id: UUID
    report_id: UUID
    schedule_id: Optional[UUID] = None
    executed_by: str
    execution_type: str
    parameters: Dict[str, Any]
    row_count: Optional[int] = None
    execution_time_ms: Optional[int] = None
    file_url: Optional[str] = None
    file_size: Optional[int] = None
    status: str
    error_message: Optional[str] = None
    executed_at: datetime
    completed_at: Optional[datetime] = None


class KPIDefinition(BaseModel):
    id: UUID
    kpi_code: str
    kpi_name: str
    category: Optional[str] = None
    calculation_sql: Optional[str] = None
    unit: str
    target_value: Optional[float] = None
    warning_threshold: Optional[float] = None
    critical_threshold: Optional[float] = None
    display_format: Optional[str] = None
    trend_direction: Optional[str] = None
    is_active: bool = True
    created_at: datetime


class KPIValue(BaseModel):
    id: UUID
    kpi_id: UUID
    value: float
    period_date: date
    period_type: str
    previous_value: Optional[float] = None
    change_percentage: Optional[float] = None
    status: Optional[str] = None
    calculated_at: datetime


class OccupancySummary(BaseModel):
    total_rooms: int
    occupied_rooms: int
    occupancy_rate: float
    arrivals_today: int
    departures_today: int
    in_house_guests: Optional[int] = None


class RevenueSummary(BaseModel):
    room_revenue: Decimal
    pos_revenue: Decimal
    other_revenue: Optional[Decimal] = None
    total_revenue: Decimal
    adr: Decimal
    rev_par: Decimal


class OperationsSummary(BaseModel):
    check_ins_completed: int
    check_outs_completed: int
    pending_arrivals: int
    pending_departures: int
    rooms_to_clean: int
    maintenance_issues: int


class DailyOperationsReport(BaseModel):
    report_date: date
    summary: Dict[str, Any]
    detailed_sections: Dict[str, Any]


class RevenueReport(BaseModel):
    summary: Dict[str, Any]
    revenue_breakdown: Dict[str, Any]
    daily_trend: Optional[List[Dict[str, Any]]] = None
    top_performers: Optional[Dict[str, Any]] = None
    year_over_year: Optional[Dict[str, Any]] = None


class ProfitLossReport(BaseModel):
    period: str
    revenue: Dict[str, Any]
    expenses: Dict[str, Any]
    profitability: Dict[str, Any]
    ratios: Dict[str, Any]
    comparison: Optional[Dict[str, Any]] = None


class GuestDemographicsReport(BaseModel):
    period: str
    total_guests: int
    demographics: Dict[str, Any]
    behavior_patterns: Dict[str, Any]
    preferences: Dict[str, Any]


class KPIDashboard(BaseModel):
    timestamp: datetime
    kpis: Dict[str, Any]
    alerts: List[Dict[str, Any]]
    period_comparison: Dict[str, Any]


class SavedFilter(BaseModel):
    id: UUID
    report_id: UUID
    user_id: UUID
    filter_name: str
    filter_config: Dict[str, Any]
    is_default: bool = False
    created_at: datetime


class ReportSubscription(BaseModel):
    id: UUID
    user_id: UUID
    report_id: UUID
    frequency: Optional[ReportPeriod] = None
    delivery_method: str
    format: ReportFormat
    custom_filters: Optional[Dict[str, Any]] = None
    is_active: bool = True
    subscribed_at: datetime


class ExecuteReportRequest(BaseModel):
    report_id: UUID
    parameters: Dict[str, Any]
    format: Optional[ReportFormat] = ReportFormat.JSON
    include_charts: Optional[bool] = False


class ExecuteReportResponse(BaseModel):
    success: bool
    data: Optional[Any] = None
    file_url: Optional[str] = None
    error: Optional[str] = None
    execution_id: Optional[UUID] = None


class CreateScheduleRequest(BaseModel):
    report_id: UUID
    schedule_name: str
    frequency: ReportPeriod
    schedule_config: Dict[str, Any]
    default_parameters: Optional[Dict[str, Any]] = None
    recipients: Dict[str, Any]
    delivery_format: ReportFormat


class OccupancyReport(BaseModel):
    period: ReportPeriod
    start_date: date
    end_date: date
    total_rooms: int
    total_available_room_nights: int
    total_occupied_room_nights: int
    occupancy_rate: float
    average_daily_rate: Decimal
    revenue_per_available_room: Decimal
    room_type_breakdown: List[Dict[str, Any]]
    daily_occupancy: List[Dict[str, Any]]


class RevenueReport(BaseModel):
    period: ReportPeriod
    start_date: date
    end_date: date
    room_revenue: Decimal
    service_revenue: Decimal
    pos_revenue: Decimal
    other_revenue: Decimal
    total_revenue: Decimal
    total_expenses: Decimal
    net_profit: Decimal
    payment_method_breakdown: Dict[str, Decimal]
    daily_revenue: List[Dict[str, Any]]
    top_revenue_sources: List[Dict[str, Any]]


class CustomerAnalytics(BaseModel):
    period: ReportPeriod
    start_date: date
    end_date: date
    new_customers: int
    returning_customers: int
    total_customers: int
    average_stay_duration: float
    average_booking_value: Decimal
    customer_satisfaction_score: Optional[float]
    top_customers: List[Dict[str, Any]]
    customer_segments: Dict[str, int]
    nationality_breakdown: Dict[str, int]


class OperationalReport(BaseModel):
    date: date
    check_ins_today: int
    check_outs_today: int
    arrivals_tomorrow: int
    departures_tomorrow: int
    rooms_occupied: int
    rooms_available: int
    rooms_under_maintenance: int
    housekeeping_status: Dict[str, int]
    pending_payments: Decimal
    tasks_pending: int
    low_stock_items: int


class FinancialSummary(BaseModel):
    period: ReportPeriod
    start_date: date
    end_date: date
    gross_revenue: Decimal
    net_revenue: Decimal
    total_expenses: Decimal
    gross_profit: Decimal
    net_profit: Decimal
    profit_margin: float
    accounts_receivable: Decimal
    accounts_payable: Decimal
    cash_flow: Decimal
    expense_breakdown: Dict[str, Decimal]


class PerformanceMetrics(BaseModel):
    period: ReportPeriod
    start_date: date
    end_date: date
    occupancy_rate: float
    adr: Decimal  # Average Daily Rate
    revpar: Decimal  # Revenue Per Available Room
    goppar: Decimal  # Gross Operating Profit Per Available Room
    average_los: float  # Length of Stay
    booking_lead_time: float
    cancellation_rate: float
    no_show_rate: float
    repeat_guest_ratio: float
    online_booking_ratio: float


class ForecastReport(BaseModel):
    period: ReportPeriod
    forecast_date: date
    projected_occupancy: float
    projected_revenue: Decimal
    projected_arrivals: int
    projected_departures: int
    room_availability: Dict[str, int]
    demand_forecast: List[Dict[str, Any]]
    revenue_forecast: List[Dict[str, Any]]
    recommendations: List[str]


class ComparisonReport(BaseModel):
    current_period: Dict[str, Any]
    previous_period: Dict[str, Any]
    year_over_year: Dict[str, Any]
    variance_analysis: Dict[str, Any]
    growth_metrics: Dict[str, float]
    trends: List[Dict[str, Any]]